### PR TITLE
Add OpenID configuration in install page

### PR DESCRIPTION
**Pull request by strk**
Allows enabling OpenID signin via installation page GUI

### CODE REVIEW

**Comment by strk** [comment]
This needs be merged before 1.2.0 is out - but there's still a problem with routes depending on the configuration and the configuration being changed by install page but not being re-read upon changing it... So it takes re-starting gitea to be able to register first user via OpenID. I've to figure out how to do the settings update from form submission

**Comment by strk** [comment]
I think the fix here is to always define the route but then return a Forbidden status if the service is not enabled

**Comment by strk** [comment]
Ready for merge, tested locally.

**Comment by strk** [comment]
@cweiske you might be interested in testing this too. It basically adds a way to configure OpenID at installation time (was missing, so it took manual intervention to enable it)

**Review by lafriks — CHANGES_REQUESTED** [review]
You have enabled all openid routes but added check that openid is enabled only in two of them

**Comment by strk** [comment]
@lafriks 754dd4ea7c9b8c009bf164153dedead98331b636 fixes that,
thanks for spotting it !

**Comment by lunny** [comment]
@strk CI failed.

File: routers/user/auth_openid.go
Comment (lafriks) [inline routers/user/auth_openid.go]: CI is failing because of space after `!` in multiple places

**Review by lafriks — CHANGES_REQUESTED** [review]

**Comment by strk** [comment]
Thanks @lafriks, I removed all those spaces and verified
that `make check` passes locally

File: routers/routes/routes.go
Comment (lafriks) [inline routers/routes/routes.go]: `ConnectOpenID` is still missing check if `EnableOpenIDSignIn` is enabled

File: routers/routes/routes.go
Comment (lafriks) [inline routers/routes/routes.go:226]: `SettingsOpenID` is still missing check if `EnableOpenIDSignIn` is enabled

File: routers/user/setting_openid.go
Comment (lafriks) [inline routers/user/setting_openid.go]: I think there should be check for `EnableOpenIDSignIn` not for `EnableOpenIDSignUp` that is for new user registration

File: routers/user/setting_openid.go
Comment (lafriks) [inline routers/user/setting_openid.go]: There should also be check for `EnableOpenIDSignIn`

File: routers/user/setting_openid.go
Comment (lafriks) [inline routers/user/setting_openid.go]: And there also should be check for `EnableOpenIDSignIn`

**Review by lafriks — CHANGES_REQUESTED** [review]

**Comment by strk** [comment]
@lafriks please see ed9ade1d
To be honest for the settings we might not even check...

File: modules/auth/user_form.go
Comment (bkcsoft) [inline modules/auth/user_form.go]: I'd rather have these both as `EnableFoo` since it makes more sense when reading

File: routers/routes/routes.go
Comment (bkcsoft) [reply routers/routes/routes.go]: Agreed, any pre-requisite settings should be checked in the route-setup, not the route-handler 🙂

**Review by bkcsoft — CHANGES_REQUESTED** [review]

**Comment by strk** [comment]
On Fri, Aug 11, 2017 at 08:59:31AM -0700, bkcsoft wrote:

> +	EnableOpenIDSignIn             bool
> +	DisableOpenIDRegistration      bool
> 
> I'd rather have these both as `EnableFoo` since it makes more sense when reading

Please take a look at the user GUI first.
There's a "DisableRegistration", so I made this
"DisableOpenIDRegistration" for consistency on that side of things.
The GUI makes the two go togheter (via Javascript).

**Comment by strk** [comment]
@bkcsoft do you really want to hold this PR back due to the enable vs. disable wording ?

**Comment by strk** [comment]
@bkcsoft I've renamed Disable to Enable, as you requested. And rebased to master.
@lafriks I think I've handled your requests too.

**Comment by strk** [comment]
actually, the change was partial, more changes were needed around, so I reverted it.

**Comment by strk** [comment]
latest commits fix the handling of checkboxes *and* turn openid sign-in on by default for new installs

**Comment by strk** [comment]
And for @bkcsoft now I've also turned the "OpenID Registration" form into an "enable" rather than "disable". Also added more javascript to check/uncheck dependents...

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
As for consideration it would probably be better as @bkcsoft recommended to create new handler/filter to check if openid signin/signup is enabled. Something like this:
https://github.com/go-gitea/gitea/blob/master/routers/routes/routes.go#L336

**Comment by strk** [comment]
Can we get this merged before 1.2.0 please ? Aesthetic changes can happen in a second step, for now it's important for the functionality to be there.

**Comment by lafriks** [comment]
I'm not against it, @bkcsoft?

**Comment by strk** [comment]
@lafriks can you give your LGTM so it's there when needed ? :)

**Comment by strk** [comment]
btw, there's something wrong with the LGTM bot, because I read `1 of 2 required approvals granted ` in the `approvals/lgtm` check but the label still reads `lgtm/need2`

**Comment by lafriks** [comment]
BTW don't write LG-TM in comments for your own PR ;) otherwise I see LG-TM 1 label

**Comment by strk** [comment]
Sorry about that, I thought it was fixed (definitely a bug, isn't it ?) - I wasn't seeing the label when I wrote the next comment, btw. Now I see it.

**Comment by lunny** [comment]
@strk it was fixed I remember.

**Comment by strk** [comment]
@lunny so it was not deployed? 
Anyway, can you review this PR  or at least mark it as targeted to 1.2?

**Comment by lunny** [comment]
@strk I don't think it's necessary for add this config to installation on v1.2 since LDAP, OAuth and SMTP didn't add to the page. And I think a middleware to check is better than so many settings check on the route functions.

**Comment by strk** [comment]
Ok now that master branch is fixed I've rebased and also committed a change using filters as suggested in previous comment. @bkcsoft can you please review and merge or at least label as targetted at 1.2.0

**Comment by lafriks** [comment]
@strk please fix formatting issues (see Drone)

**Comment by strk** [comment]
@lunny I committed the middleware check and dropped the checks in the
route functions. As per installation steps: OpenID is special in that
it is a config file, not database, configuration. That's why LDAP
and OAuth and SMTP do not belong in the installation page, while
OpenID does.

**Comment by lafriks** [comment]
LG-TM

**Review by bkcsoft — APPROVED** [review]

**Comment by bkcsoft** [comment]
LGTM

**Pull request merged**
